Improvement/s3 utils 171 bump deps

benzekrimaha opened the pull request:
Issue: S3UTILS-171

bert-e commented:
# Hello benzekrimaha,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/s3utils/doc/user).


<details>
  <summary><b>Available options</b></summary>

  name   | description  | privileged   | authored
  ------ | ------------ | ------------ |----------
  `/after_pull_request` | Wait for the given pull request id to be merged before continuing with the current one. |  | 
  `/bypass_author_approval` | Bypass the pull request author's approval |  :star:  | 
  `/bypass_build_status` | Bypass the build and test status |  :star:  | 
  `/bypass_commit_size` | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
  `/bypass_incompatible_branch` | Bypass the check on the source branch prefix |  :star:  | 
  `/bypass_jira_check` | Bypass the Jira issue check |  :star:  | 
  `/bypass_peer_approval` | Bypass the pull request peers' approval |  :star:  | 
  `/bypass_leader_approval` | Bypass the pull request leaders' approval |  :star:  | 
  `/approve` | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
  `/create_pull_requests` | Allow the creation of integration pull requests. |  | 
  `/create_integration_branches` | Allow the creation of integration branches. |  | 
  `/no_octopus` | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
  `/unanimity` | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
  `/wait` | Instruct Bert-E not to run until further notice. |  | 
  

</details>



<details>
  <summary><b>Available commands</b></summary>

  name   | description  | privileged
  ------ | ------------ | ------------
  `/help` | Print Bert-E's manual in the pull request. | 
  `/status` | Print Bert-E's current status in the pull request ```TBA``` | 
  `/clear` | Remove all comments from Bert-E from the history ```TBA``` | 
  `/retry` | Re-start a fresh build ```TBA``` | 
  `/build` | Re-start a fresh build ```TBA``` | 
  `/force_reset` | Delete integration branches & pull requests, and restart merge process from the beginning. | 
  `/reset` | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
  
</details>




*Status report is not available.*

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-171 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.14.19`

* `1.15.9`


Please check the `Fix Version/s` of S3UTILS-171, or the target
branch of this pull request.

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-171 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.15.9`


Please check the `Fix Version/s` of S3UTILS-171, or the target
branch of this pull request.

benzekrimaha committed:
bump node version and dependencies

Issue: S3UTILS-171

benzekrimaha committed:
Update eslint configuration

Issue: S3UTILS-171

benzekrimaha committed:
fixup on files post lint config update

Issue: S3UTILS-171

benzekrimaha committed:
Tests and code adjustments post bumps

- The async module bump lead to the need of using callbacks on functions like
doWhilst.
- Also we need to add for mongo some parameters to have the expected
return , in this specific case  returnDocument: 'after',
includeResultMetadata: true.
- In our tests, bucketInfos used to send an empty array,
this commit actually uses fill to send the intended array as before
async used to just ignore the empty array and return the expected result.

Issue: S3UTILS-171

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-ARG NODE_VERSION=16.20.2-bullseye-slim
+ARG NODE_VERSION=22-bullseye-slim
```
should use bookworm & specify the exact patch version to use (we don't want to have un-controlled version bumps)

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue S3UTILS-171 contains:


* `1.16.0`



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `1.15.9`


Please check the `Fix Version/s` of S3UTILS-171, or the target
branch of this pull request.

francoisferrand commented:
this PR should target a new branch, to ensure we don't introduce breaking change

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on CRR/ReplicationStatusUpdater.js:
```diff
@@ -352,7 +352,7 @@ class ReplicationStatusUpdater {
                 },
             ),
-            () => {
+            callback => {
```
best to just make the function `async` :


```suggestion
            async () => {
```

francoisferrand commented on CompareRaftMembers/BlockDigestsStorage.js:
```diff
@@ -38,10 +38,10 @@ class BlockDigestsStorage extends stream.Writable {
+        const { size, digest, lastKey } = blockInfo;
         this.cargo.push({
             type: 'put',
-            key: blockInfo.lastKey, // index by last block key for efficient lookup
```
we should keep the comment, no reason to remove it?

francoisferrand commented on CountItems/CountManager.js:
```diff
@@ -167,7 +167,17 @@ class CountManager {
-        this.q.push(bucketInfos);
+        bucketInfos.forEach(bucketInfo => {
+            this.q.push({bucketInfo}, err => {
+                if (err) {
+                    this.log.error('error processing bucket', {
+                        error: err,
+                        method: 'CountManager::addWork',
+                    });
+                }
+            }
```
should we not just add this log in the "global" error handler (L.201), instead of adding one hadnler per task?

```
this.q.error(err => {
    this.log.error('error processing bucket', {
        error: err,
        method: 'CountManager::addWork',
    });
    this.q.pause();
    this.q.kill();
    return process.nextTick(onceCB, err);
});
```

and keep using a single push
```suggestion
this.q.push(bucketInfos);
```

francoisferrand commented on CountItems/CountWorkerObj.js:
```diff
@@ -21,9 +21,9 @@ class CountWorkerObj {
         });
         this._worker.on('message', data => {
-            if (data.owner !== 'scality') return;
+            if (data.owner !== 'scality') {return;}
```
```suggestion
            if (data.owner !== 'scality') {
                return;
            }
```

francoisferrand commented on CountItems/CountWorkerObj.js:
```diff
@@ -21,9 +21,9 @@ class CountWorkerObj {
+            if (data.owner !== 'scality') {return;}
             const cb = this._getCallback(data.id);
-            if (!cb) return;
+            if (!cb) {return;}
```
same

francoisferrand commented on StalledRetry/StalledRequestHandler.js:
```diff
@@ -17,7 +17,9 @@ class StalledRequestHandler {
 
     queueSetup() {
-        this._queue = async.queue(({ bucket, batch, getNext }, done) => {
+        this._queue = async.queue((task, done) => {
```
why the change? is it not working anymore, or flagged by eslint?

benzekrimaha replied:
not working anymore with the async bump

francoisferrand replied:
that is weird, should be purely a node "syntax" thing, unrelated to async :-/

francoisferrand commented on StalledRetry/StalledRequestHandler.js:
```diff
@@ -68,9 +71,11 @@ class StalledRequestHandler {
-            () => (!this.killed && this.isInProgress()),
-            done => setTimeout(done, 1000),
-            cb,
+            done => {
+                done(null, !this.killed && this.isInProgress());
+            },
+            next => setTimeout(next, 1000),
```
just use async and keep code leaner:

```suggestion
            async () => (!this.killed && this.isInProgress()),
            done => setTimeout(done, 1000),
```

francoisferrand commented on VerifyReplication/verifyReplication.js:
```diff
@@ -88,8 +88,7 @@ function handlePrefixes(prefixList, cb) {
             listingLimit,
-        };
-        // eslint-disable-next-line no-use-before-define
+        };       
```
extra space at EOL

francoisferrand commented on compareBuckets/compareBuckets.js:
```diff
@@ -91,15 +87,14 @@ function compareBuckets(params, log, cb) {
                             }
-
-                            srcContents = contents;
+                            srcContents = contents || [];
```
why this chnage?
contents cannot really be empty... esp. looking at `listBucketMasterKeys()`, we expect it to be returned, and access elements within anyway: so it would fail (crash) if no present.

francoisferrand commented on compareBuckets/compareBuckets.js:
```diff
@@ -116,10 +111,10 @@ function compareBuckets(params, log, cb) {
                             }
-
-                            dstContents = contents;
+                            dstContents = contents || [];
```
same

francoisferrand commented on compareBuckets/compareBuckets.js:
```diff
@@ -224,15 +220,16 @@ function compareBuckets(params, log, cb) {
             });
         },
-        () => (!srcDone || !dstDone || srcContents.length > 0 || dstContents.length > 0),
+        callback => {
+            callback(null, !srcDone || !dstDone || srcContents.length > 0 || dstContents.length > 0);
+        },
         err => {
```
```suggestion
        async () => (!srcDone || !dstDone || srcContents.length > 0 || dstContents.length > 0),
        err => {
```

francoisferrand commented on service-level-sidecar/bucketd.js:
```diff
@@ -37,8 +37,7 @@ async function* listBuckets(log) {
 
-        try {
-            // eslint-disable-next-line no-await-in-loop
+        try {             
```
spaces at EOL

francoisferrand commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-ARG NODE_VERSION=16.20.2-bullseye-slim
+ARG NODE_VERSION=22.14.0-bookworm
```
should still use the `-slim` variant
```suggestion
ARG NODE_VERSION=22.14.0-bookworm-slim
```

francoisferrand commented on Dockerfile:
```diff
@@ -1,4 +1,4 @@
-ARG NODE_VERSION=16.20.2-bullseye-slim
+ARG NODE_VERSION=22.14.0-bookworm
 
```
At the moment, the image includes python, for the scripts in the `utapi` folder (c.f. https://scality.atlassian.net/browse/S3UTILS-172)

With the upgrade, a newer version of python will be used: so we need to check that the scripts still work (or find a way to remove them and drop python from the image, i.e. move forward with s3utils-172...)

francoisferrand replied:
there is now `22.15.0-bookworm-slim`, should we use it?
or just "plan" a global bump from 22.14 to 22.15 ?

francoisferrand commented on Dockerfile:
```diff
@@ -34,12 +34,14 @@ RUN apt-get update && \
 ENV BALLOT_VERSION 1.0.4
 ADD https://github.com/scality/ballot/releases/download/v${BALLOT_VERSION}/ballot-v${BALLOT_VERSION}-linux-amd64 /usr/src/app/ballot
 RUN chmod +x /usr/src/app/ballot
 
+
```
extra empty line

williamlardier commented on CountItems/CountManager.js:
```diff
@@ -188,18 +198,18 @@ class CountManager {
             if (this.q.idle()) {
                 this.q.pause();
-                this.q.kill();
+                this.q.kill(); 
```
```suggestion
                this.q.kill();
```

williamlardier commented on StalledRetry/StalledRequestHandler.js:
```diff
@@ -68,9 +71,9 @@ class StalledRequestHandler {
+            async () => (!this.killed && this.isInProgress()),
             done => setTimeout(done, 1000),
-            cb,
+            cb
```
```suggestion
            cb,
```

williamlardier commented on StalledRetry/StalledRequestHandler.js:
```diff
@@ -68,9 +71,9 @@ class StalledRequestHandler {
     _waitForCompletion(cb) {
         async.whilst(
-            () => (!this.killed && this.isInProgress()),
+            async () => (!this.killed && this.isInProgress()),
```
```suggestion
            () => (!this.killed && this.isInProgress()),
```

williamlardier commented on compareBuckets/compareBuckets.js:
```diff
@@ -224,15 +220,14 @@ function compareBuckets(params, log, cb) {
             });
         },
-        () => (!srcDone || !dstDone || srcContents.length > 0 || dstContents.length > 0),
+        async () => (!srcDone || !dstDone || srcContents.length > 0 || dstContents.length > 0),
```
```suggestion
        () => (!srcDone || !dstDone || srcContents.length > 0 || dstContents.length > 0),
```

williamlardier replied:
(same at other places)
There is no need to specify the async keywork if the function is not asynchronous?

benzekrimaha replied:
Without async, the function returns a raw boolean, and after the async bump the  async.whilst() doesn't handle the boolean return value as it's expecting a promise this is why the async is used.

tcarmet commented on package.json:
```diff
@@ -26,22 +26,22 @@
-    "aws-sdk": "^2.1005.0",
-    "bucketclient": "git+https://github.com/scality/bucketclient#8.1.5",
-    "express": "^4.18.2",
+    "arsenal": "git+https://github.com/scality/arsenal#8.2.1",
```
arsenal and bucketclient got a couple more updates, I'll leave it up to you to check if you want to take the opportunity to make a bump now.

benzekrimaha replied:
Indeed , I bumped both to the latest versions

tcarmet commented on eslint.config.mjs:
```diff
@@ -0,0 +1,47 @@
+            "no-redeclare":"off",
+        }
+    }
+];
```
```suggestion
];

```

williamlardier commented on tests/constants.js:
```diff
@@ -29,6 +29,36 @@ module.exports = {
         },
         ingestion: null,
     },
+    InternalTestBucketMD: {
```
```suggestion
    internalTestBucketMD: {
```

williamlardier reviewed (approved):
You can add a commit to bump if you want :slightly_smiling_face:

ghost commented:
There is also a unit test failure that requires attention, as it affects a sub dependency...

benzekrimaha commented:
/approve

francoisferrand commented on CountItems/CountWorker.js:
```diff
@@ -36,7 +36,7 @@ class CountWorker {
             Object.assign(bucketInfoObj, { _websiteConfiguration: null });
         }
-        const bucketInfo = BucketInfo.fromObj(bucketInfoObj);
+        const bucketInfo = BucketInfo.fromObj(bucketInfoObj.bucketInfo ? bucketInfoObj.bucketInfo : bucketInfoObj);
```
```suggestion
        const bucketInfo = BucketInfo.fromObj(bucketInfoObj.bucketInfo || bucketInfoObj);
```

francoisferrand commented on compareBuckets/compareBuckets.js:
```diff
@@ -224,15 +220,14 @@ function compareBuckets(params, log, cb) {
             statusObj.dstBucketInProgress = null;
             statusObj.srcBucketInProgress = null;
-            cb(err);
+            cb(err, statusObj);
```
callback "result" is not used (only the first param, err), so why do we need to add it?

* this variable is actually "shared" which the caller, who provides it when calling the function:
  ```
  compareBuckets(params, log, err => {
  ```

* and the code cannot so easily be cleaned' up to "return" the object, as it is used during the whole processing, to display process:
   ```
   setInterval(() => logProgress('progress update', status), LOG_PROGRESS_INTERVAL * 1000);
   ```

benzekrimaha commented on package.json:
```diff
@@ -26,22 +26,22 @@
-    "aws-sdk": "^2.1005.0",
-    "bucketclient": "git+https://github.com/scality/bucketclient#8.1.5",
-    "express": "^4.18.2",
+    "arsenal": "git+https://github.com/scality/arsenal#f738d26a9084dab707cbc3abfd9bcdd6de7e73d2",
```
Please note that this hash will be changed to the accurate version once the arsenal PR merged.

benzekrimaha commented:
/approve

benzekrimaha commented:
/wait

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.16`



The following branches have **NOT** changed:

* `development/1.13`
* `development/1.14`
* `development/1.15`
* `development/1.4`



Please check the status of the associated issue S3UTILS-171.

Goodbye benzekrimaha.




*The following options are set:* **approve**
